extensions/interactions/CodeRepl/CodeRepl.js

```diff
@@ -114,27 +128,65 @@ oppia.directive('oppiaInteractiveCodeRepl', [
           });
         };
 
-        var fixLineNumbers = function(err) {
-          var preCodeNumLines = $scope.preCode.split('\n').length;
-          var userCodeNumLines = $scope.code.split('\n').length;
-          return err.replace(/on line ([0-9]+)/g, function(match, p1) {
-            var originalLineNumber = Number(p1);
-
-            return (
-              originalLineNumber <= preCodeNumLines ?
-              '(on line ' + String(originalLineNumber) +
-                ' of exploration creator\'s prepended code)' :
-              originalLineNumber <= preCodeNumLines + userCodeNumLines ?
-              '(on line ' + String(originalLineNumber - preCodeNumLines) + ')' :
-              '(on line ' + String(
-                  originalLineNumber - preCodeNumLines - userCodeNumLines) +
-                ' of exploration creator\'s appended code)');
-          });
+        var initMarkers = function(editor) {
+          var doc = editor.getDoc();
+
+          // The -1 here is because prepended code ends with a newline.
+          var preCodeNumLines = $scope.preCode.split('\n').length - 1;
+          var postCodeNumLines = $scope.postCode.split('\n').length;
+          var fullCodeNumLines = $scope.code.split('\n').length;
+          var userCodeNumLines = (
+              fullCodeNumLines - preCodeNumLines - postCodeNumLines);
+
+          // Mark pre- and post- code as uneditable, and give it some styling.
+          var markOptions = {
+            atomic: true,
+            readOnly: true,
+            inclusiveLeft: true,
+            inclusiveRight: true
+          };
+
+          if ($scope.preCode.length !== 0) {
+            doc.markText(
+                {
```

Comment:
Optional nit: we generally indent everything by 2 in JS, including continuation lines.